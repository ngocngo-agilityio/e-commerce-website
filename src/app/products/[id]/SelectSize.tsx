'use client';

// Libs
import { memo } from 'react';
import isEqual from 'react-fast-compare';

// Types
import { SizeOption } from '@types';

// Components
import { SelectSize as SelectSizeComponent } from '@components';

interface Props {
  sizes: SizeOption[];
}

const SelectSize = ({ sizes }: Props): JSX.Element => {
  return <SelectSizeComponent options={sizes} onChange={() => {}} />;
};

export default memo(SelectSize, isEqual);
