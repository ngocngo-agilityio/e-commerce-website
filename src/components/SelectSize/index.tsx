'use client';

// Libs
import { useCallback, MouseEvent, useState, memo } from 'react';
import isEqual from 'react-fast-compare';
import {
  Button,
  Flex,
  Menu,
  MenuButton,
  MenuItem,
  MenuList,
  Text,
} from '@chakra-ui/react';

// Assets
import { AngleDownIcon } from '@assets';

// Types
import { SizeOption } from '@types';

interface Props {
  options: SizeOption[];
  onChange: (selectedOption: SizeOption | undefined) => void;
}

const SelectSize = ({ options = [], onChange }: Props): JSX.Element => {
  const [selectedOption, setSelectedOption] = useState<SizeOption>();

  const handleOnChange = useCallback(
    (e: MouseEvent<HTMLButtonElement>) => {
      const value = e.currentTarget.value;
      const selectedOption = options.find((item) => item.value === value);

      setSelectedOption(selectedOption);
      onChange(selectedOption);
    },
    [onChange, options],
  );

  return (
    <Menu size="sm">
      <MenuButton as={Button} variant="selectBtn" rightIcon={<AngleDownIcon />}>
        {selectedOption?.label || 'Select Size'}
      </MenuButton>
      <MenuList>
        {options.map((item) => {
          const { value, label, code } = item || {};

          return (
            <MenuItem
              key={value}
              value={value}
              borderBottom="1px"
              borderColor="menu.border"
              onClick={handleOnChange}
            >
              <Flex justifyContent="space-between" w="full">
                <Text color="menu.text">{label}</Text>
                <Text color="menu.text">{code}</Text>
              </Flex>
            </MenuItem>
          );
        })}
      </MenuList>
    </Menu>
  );
};

export default memo(SelectSize, isEqual);
