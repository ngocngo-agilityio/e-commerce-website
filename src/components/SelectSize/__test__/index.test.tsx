// Libs
import { screen, render, fireEvent } from '@testing-library/react';

// Components
import SelectSize from '..';

const mockProps = {
  options: [
    { value: '1', label: 'Small', code: 'S' },
    { value: '2', label: 'Medium', code: 'M' },
    { value: '3', label: 'Large', code: 'L' },
    { value: '4', label: 'Extra Large', code: 'XL' },
  ],
  onChange: jest.fn(),
};

describe('SelectSize component', () => {
  test('should render SelectSize successfully', () => {
    render(<SelectSize {...mockProps} />);

    expect(screen.getByText(mockProps.options[0].label)).toBeInTheDocument();
  });

  test('should match snapshot for SelectSize', () => {
    const { container } = render(<SelectSize {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('should call onChange when select the size option', () => {
    render(<SelectSize {...mockProps} />);

    fireEvent.click(screen.getByText(mockProps.options[1].label));

    expect(mockProps.onChange).toHaveBeenCalled();
  });
});
