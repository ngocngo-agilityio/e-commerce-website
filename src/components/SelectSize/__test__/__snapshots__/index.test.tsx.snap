// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectSize component should match snapshot for SelectSize 1`] = `
<div>
  <button
    aria-controls="menu-list-:r7:"
    aria-expanded="false"
    aria-haspopup="menu"
    class="chakra-button chakra-menu__menu-button css-4xx2wk"
    id="menu-button-:r7:"
    type="button"
  >
    <span
      class="css-xl71ch"
    >
      Select Size
    </span>
    <span
      class="chakra-button__icon css-1hzyiq5"
    >
      <svg
        class="chakra-icon css-1mvevqt"
        focusable="false"
        viewBox="0 0 13 8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5.66016 7.69531C5.90625 7.94141 6.31641 7.94141 6.5625 7.69531L11.8945 2.39062C12.1406 2.11719 12.1406 1.70703 11.8945 1.46094L11.2656 0.832031C11.0195 0.585938 10.6094 0.585938 10.3359 0.832031L6.125 5.04297L1.88672 0.832031C1.61328 0.585938 1.20312 0.585938 0.957031 0.832031L0.328125 1.46094C0.0820312 1.70703 0.0820312 2.11719 0.328125 2.39062L5.66016 7.69531Z"
          fill="#818181"
        />
      </svg>
    </span>
  </button>
  <div
    class="css-0"
    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-sjt5nk"
      id="menu-list-:r7:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8);"
      tabindex="-1"
    >
      <button
        class="chakra-menu__menuitem css-1edjah3"
        data-index="0"
        id="menu-list-:r7:-menuitem-:r8:"
        role="menuitem"
        tabindex="-1"
        type="button"
        value="1"
      >
        <div
          class="css-l64258"
        >
          <p
            class="chakra-text css-9ooqc1"
          >
            Small
          </p>
          <p
            class="chakra-text css-9ooqc1"
          >
            S
          </p>
        </div>
      </button>
      <button
        class="chakra-menu__menuitem css-1edjah3"
        data-index="1"
        id="menu-list-:r7:-menuitem-:r9:"
        role="menuitem"
        tabindex="-1"
        type="button"
        value="2"
      >
        <div
          class="css-l64258"
        >
          <p
            class="chakra-text css-9ooqc1"
          >
            Medium
          </p>
          <p
            class="chakra-text css-9ooqc1"
          >
            M
          </p>
        </div>
      </button>
      <button
        class="chakra-menu__menuitem css-1edjah3"
        data-index="2"
        id="menu-list-:r7:-menuitem-:ra:"
        role="menuitem"
        tabindex="-1"
        type="button"
        value="3"
      >
        <div
          class="css-l64258"
        >
          <p
            class="chakra-text css-9ooqc1"
          >
            Large
          </p>
          <p
            class="chakra-text css-9ooqc1"
          >
            L
          </p>
        </div>
      </button>
      <button
        class="chakra-menu__menuitem css-1edjah3"
        data-index="3"
        id="menu-list-:r7:-menuitem-:rb:"
        role="menuitem"
        tabindex="-1"
        type="button"
        value="4"
      >
        <div
          class="css-l64258"
        >
          <p
            class="chakra-text css-9ooqc1"
          >
            Extra Large
          </p>
          <p
            class="chakra-text css-9ooqc1"
          >
            XL
          </p>
        </div>
      </button>
    </div>
  </div>
</div>
`;
